### PR TITLE
improve: add common class extension to SDKUtils

### PR DESCRIPTION
**Pull request by MdUsmanGani**
# Description
Add common class extension to SDKUtils

# Checklist
- [x] I have read the [contributing guidelines](../CONTRIBUTING.md).
- [x] I wrote/updated tests for new/changed code
- [x] I removed all sensitive data **before every commit**, including API endpoints and keys
- [x] I checked that all possible internal and platform/library exceptions are gracefully handled and **will not** crash the host app
- [x] I checked for open & known issues (especially crashes) for any newly added platform APIs or libraries
- [x] I tested non-trivial changes on a real device
- [x] I ran `./gradlew check` without errors

### CODE REVIEW

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    }
+
+    /**
+     * Check whether the strings are same.
+     *
+     * @param first the first string
+     * @param second the second string
+     * @return true if same otherwise false
+     */
+    fun isEqual(first: String, second: String): Boolean {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: No need for this. This is already available for String classes.

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        return first == second
+    }
+
+    /**
+     * Check the string whether null or empty
+     *
+     * @param rawString the raw string
+     * @return true if null or empty otherwise false
+     */
+    fun isNullOrEmpty(rawString: String?): Boolean {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: No need for this. This is already available for String classes.

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        return true
+    }
+
+    /**
+     * Get the UTC date format from the simple date format, i.e. "yyyy-MM-dd'T'HH:mm:ssX"
+     *
+     * @return the date format
+     */
+    @SuppressLint("SimpleDateFormat")
+    fun getUTCDateFormat(): DateFormat {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt:16]: Can we make this as extension of `DateFormat`?

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    private val logger = Logger(CommonUtil::class.java.simpleName)
+
+    /**
+     * Get the encoded byte array of string from character set
+     *
+     * @param rawString the raw string.
+     * @param charsetName the character set, i.e. "UTF-8", "UTF-16"
+     * @return the ByteArray of the string
+     */
+    fun getEncodedByteArray(rawString: String, charsetName: String): ByteArray {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Can we make this as extension of `String` class? and separate UTF-8 and UTF-16 functions.

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        return rawString.toByteArray(Charset.forName(charsetName))
+    }
+
+    /**
+     * Get the encoded MD5 digest string
+     *
+     * @param rawString the raw string.
+     * @return the MD5 digest string
+     */
+    fun getMD5HashData(rawString: String?): String? {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Can we make this as extension of `String` class?

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        return rawString.isNullOrEmpty()
+    }
+
+    /**
+     * Get the encoded Sha256 digest string
+     *
+     * @param rawString the raw string.
+     * @return the encoded string
+     */
+    fun getSha256HashData(rawString: String?): String? {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Can we make this as extension of `String` class?

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+     * Check the device to make sure it has the Google Play Services APK.
+     *
+     *
+     * If it doesn't, display a dialog that allows users to download the APK from the Google Play
+     * Store or enable it in the device's system settings.
+     *
+     * @param context non null activity, used to display update dialog if necessary.
+     * @return true if play services are available, false otherwise
+     */
+    fun checkPlayServices(context: Activity): Boolean {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Can we make this as extension of `Context` class?

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    }
+
+    /**
+     * Get the UTC date from a sample date string
+     *
+     * @param dateString the date string.
+     * @return the date object based on the date string
+     */
+    @SuppressLint("SimpleDateFormat")
+    fun getUTCDate(dateString: String?): Date? {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Can we make this as extension of `Date` class?

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+    }
+
+    /**
+     * Get the resource id from drawable folder.
+     *
+     * @param context the application context.
+     * @param resourceName the resource name
+     * @return the drawable resource id
+     */
+    fun getDrawableResourceId(context: Context, resourceName: String?): Int {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Can we make this as extension of `Context` class?

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+    }
+
+    /**
+     * Get the resource id from raw folder.
+     *
+     * @param context the application context.
+     * @param resourceName the resource name
+     * @return the raw resource id
+     */
+    fun getRawResourceId(context: Context, resourceName: String?): Int {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Can we make this as extension of `Context` class?

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+    }
+
+    /**
+     * Parse the color string, and return the corresponding color-int.
+     *
+     * @param colorString the color string.
+     * @return the color int value
+     */
+    @ColorInt
+    fun getColorValue(colorString: String?): Int {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Can we make this as extension of `String` class?

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+    }
+
+    /**
+     * Get the desired file path exists in assets folder.
+     *
+     * @param context the application context.
+     * @param fileName the file name
+     * @return the asset path of the file
+     */
+    fun getAssetsFilePath(context: Context, fileName: String?): String? {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Can we make this as extension of `Context` class?

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+        return null
+    }
+
+    /**
+     * Get the device mode (Dark or Light)
+     *
+     * @param context the application context.
+     * @return true if dark mode otherwise false for light mode
+     */
+    fun isDarkMode(context: Context): Boolean {
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Can we make this as extension of `Context` class?

**Review by vincentdelmundo — CHANGES_REQUESTED** [review]
Looks good. Please refactor the functions to separate extension classes.

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 2 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: I also thought this is not required. Removed. Thanks.

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 1 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Removed.

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 5 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Done

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 9 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Done

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 10 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Done

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 11 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: changed

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 13 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: updated

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 12 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: updated

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 4 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: updated

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 3 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt:16]: I could not make it as DateFormat is an abstract java class. Any help?

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 6 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: updated

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 7 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: updated

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 8 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: I tried but not working, please give some idea/example

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+     * @return the MD5 digest string
+     */
+    fun String.getMD5HashData(): String? {
+        return try {
+            String.format(
+                "%032x",
+                BigInteger(
+                    1,
+                    MessageDigest.getInstance("MD5")
+                        .digest(getUTF8ByteArray())
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: This need `this. getUTF8ByteArray`

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+
+    /**
+     * Get the encoded Sha256 digest string
+     *
+     * @return the encoded string
+     */
+    fun String.getSha256HashData(): String? {
+        return try {
+            val md = MessageDigest.getInstance("SHA-256")
+            val byteData = md.digest(getUTF8ByteArray())
```
Comment (vincentdelmundo) [inline sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: This need `this. getUTF8ByteArray`

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 3 above)
Comment (vincentdelmundo) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt:16]: I think this is fine.

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 8 above)
Comment (vincentdelmundo) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: Better to move this to `String` extension

File: README.md [hunk 16]
```diff
@@ -123,6 +123,24 @@ val path = "my-file.json"
 val device: Device? = gson.fromResources(path, Device::class.java)
 ```
 
+### <a name="#common-usage"></a>Common Utility
+Please check the following sample code to use the common utility function (s).
```
Comment (vincentdelmundo) [inline README.md]: Change example to using extension classes.

**Review by vincentdelmundo — CHANGES_REQUESTED** [review]
Added some comments. Let's move each extension function to separate classes like StringExtension.kt

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 14 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: update

File: sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt (hunk 15 above)
Comment (MdUsmanGani) [reply sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/CommonUtil.kt]: updated

File: README.md (hunk 16 above)
Comment (MdUsmanGani) [reply README.md]: updated

**Review by vincentdelmundo — APPROVED** [review]
LGTM. Please improved code coverage.

**Pull request merged**